contracts/governance/Vesting/VestingLogic.sol

```diff
@@ -110,68 +120,79 @@ contract VestingLogic is IVesting, VestingStorage, ApprovalReceiver {
-        _withdrawTokens(receiver, false);
+    function withdrawTokensStartingFrom(address _receiver, uint256 _startFrom)
+        external
+        onlyOwners
+    {
+        _withdrawTokens(_receiver, _startFrom);
     }
 
     /**
      * @notice Withdraws tokens from the staking contract and forwards them
      * to an address specified by the token owner. Low level function.
      * @dev Once here the caller permission is taken for granted.
-     * @param receiver The receiving address.
-     * @param isGovernance Whether all tokens (true)
+     * @param _receiver The receiving address.
+     * @param _startFrom The start value for the iterations.
      * or just unlocked tokens (false).
      * */
-    function _withdrawTokens(address receiver, bool isGovernance) internal {
-        require(receiver != address(0), "receiver address invalid");
+    function _withdrawTokens(address _receiver, uint256 _startFrom) internal {
+        require(_receiver != address(0), "receiver address invalid");
 
         uint96 stake;
 
         /// @dev Usually we just need to iterate over the possible dates until now.
         uint256 end;
 
+        uint256 defaultStart = startDate + cliff;
+
+        _startFrom = _startFrom >= defaultStart ? _startFrom : defaultStart;
+
         /// @dev In the unlikely case that all tokens have been unlocked early,
         ///   allow to withdraw all of them.
-        if (staking.allUnlocked() || isGovernance) {
+        if (staking.allUnlocked()) {
             end = endDate;
         } else {
             end = block.timestamp;
         }
 
+        /// @dev max iterations need to be decreased by 1, otherwise the iteration will always be surplus by 1
+        uint256 totalIterationValue =
+            (_startFrom + (FOUR_WEEKS * (getMaxVestingWithdrawIterations() - 1)));
+        uint256 adjustedEnd = end < totalIterationValue ? end : totalIterationValue;
+
         /// @dev Withdraw for each unlocked position.
         /// @dev Don't change FOUR_WEEKS to TWO_WEEKS, a lot of vestings already deployed with FOUR_WEEKS
         ///		workaround found, but it doesn't work with TWO_WEEKS
-        for (uint256 i = startDate + cliff; i <= end; i += FOUR_WEEKS) {
+        for (uint256 i = _startFrom; i <= adjustedEnd; i += FOUR_WEEKS) {
             /// @dev Read amount to withdraw.
             stake = staking.getPriorUserStakeByDate(address(this), i, block.number - 1);
 
             /// @dev Withdraw if > 0
             if (stake > 0) {
-                if (isGovernance) {
-                    staking.governanceWithdraw(stake, i, receiver);
-                } else {
-                    staking.withdraw(stake, i, receiver);
-                }
+                staking.withdraw(stake, i, _receiver);
             }
         }
 
-        emit TokensWithdrawn(msg.sender, receiver);
+        if (adjustedEnd < end) {
+            emit TeamVestingPartiallyCancelled(msg.sender, _receiver, adjustedEnd);
```

Comment:
this function is also called to withdraw unlocked tokens, in which case this event is irrelevant  
besides, i believe it was decided to cancel vesting contracts directly only, i.e. in the Staking contract - i believe cancellation from VestingLogic  should have been removed why is the cancellation in the VestingLogic contract, ain't it?

Reply:
There might be a case where the user will reach the limit of maxIteration when trying to withdraw. This event is used to indicate to the users about the last processed state. Maybe we can find a better event name here?

Reply:
@tjcloa is right here. this function is used for regular token withdrawals from the vesting contract. it was in the past also used by the governanceWithdraw function, but this was removed, so now it's exclusively used for regular withdrawals. therefore the event is simply wrong. It should still be the old TokensWithdrawn event. Also to not drive betsy and remy crazy. remember that event changes are a big deal for them.

i think it's questionable if we should update the vesting logic here at all. i think, we can leave it unchanged. 
we might want to introduce a better vesting logic at some point, so it would be good to extract the removal of the governanceWithdraw function to a separate PR. it's also a good idea to change the regular withdraw function to work with partial withdraws, but again, i do not think it has to be in this PR. let's focus on the staking and vesting registry changes here, please.

Reply:
also remember that we can't update the existing vesting contracts. so if we run into the situation where users can't withdraw their tokens anymore, we'll likely need to introduce another workaround on the staking contract, sad as it is.

Reply:
Vesting logic & four year vesting logic has been reverted in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/447/commits/21dd284dc55c0b0355bb165ce2e24dbf154e32e6)

Reply:
created the relevant `VestingLogic` update story